interfaces/builtin/dbus.go

```diff
@@ -0,0 +1,426 @@
+dbus (receive, send)
+    bus=###DBUS_BUS###
+    interface=###DBUS_INTERFACE###
+    peer=(label=###SLOT_SECURITY_TAGS###),
+
+# allow connected snaps to all interfaces via ###DBUS_PATH### (eg,
+# org.freedesktop.*, org.gtk.Application, etc) to allow full integration with
+# connected snaps.
+dbus (receive, send)
+    bus=###DBUS_BUS###
+    path=###DBUS_PATH###
+    peer=(label=###SLOT_SECURITY_TAGS###),
+`
+
+const dbusConnectedPlugSecComp = `
+getsockname
+recvmsg
+sendmsg
+sendto
+`
+
+type DbusInterface struct{}
+
+func (iface *DbusInterface) Name() string {
+	return "dbus"
+}
+
+// Obtain yaml-specified bus well-known name
+func (iface *DbusInterface) getAttribs(attribs map[string]interface{}) (string, string, error) {
+	// bus attribute
+	bus, ok := attribs["bus"].(string)
+	if !ok {
+		return "", "", fmt.Errorf("cannot find attribute 'bus'")
+	}
+
+	if bus != "session" && bus != "system" {
+		return "", "", fmt.Errorf("bus '%s' must be one of 'session' or 'system'", bus)
+	}
+
+	// name attribute
+	name, ok := attribs["name"].(string)
+	if !ok {
+		return "", "", fmt.Errorf("cannot find attribute 'name'")
+	}
+
+	err := interfaces.ValidateDBusBusName(name)
+	if err != nil {
+		return "", "", err
+	}
+
+	// snapd has AppArmor rules (see above) allowing binds to busName-PID
+	// so to avoid overlap with different snaps (eg, busName running as PID
+	// 123 and busName-123), don't allow busName to end with -PID. If that
+	// rule is removed, this limitation can be lifted.
+	invalidSnappyBusName := regexp.MustCompile("-[0-9]+$")
+	if invalidSnappyBusName.MatchString(name) {
+		return "", "", fmt.Errorf("DBus bus name must not end with -NUMBER")
+	}
+
+	return bus, name, nil
+}
+
+// Determine AppArmor dbus abstraction to use based on bus
+func getAppArmorAbstraction(bus string) (string, error) {
+	var abstraction string
+	if bus == "system" {
+		abstraction = "dbus-strict"
+	} else if bus == "session" {
+		abstraction = "dbus-session-strict"
+	} else {
+		return "", fmt.Errorf("unknown abstraction for specified bus '%q'", bus)
+	}
+	return abstraction, nil
+}
+
+// Calculate individual snippet policy based on bus and name
+func getAppArmorSnippet(policy []byte, bus string, name string) []byte {
+	old := []byte("###DBUS_BUS###")
+	new := []byte(bus)
+	snippet := bytes.Replace(policy, old, new, -1)
+
+	old = []byte("###DBUS_NAME###")
+	new = []byte(name)
+	snippet = bytes.Replace(snippet, old, new, -1)
+
+	// convert name to AppArmor dbus path (eg 'org.foo' to '/org/foo{,/**}')
+	var pathBuf bytes.Buffer
+	pathBuf.WriteString(`"/`)
+	pathBuf.WriteString(strings.Replace(name, ".", "/", -1))
+	pathBuf.WriteString(`{,/**}"`)
+
+	old = []byte("###DBUS_PATH###")
+	new = pathBuf.Bytes()
+	snippet = bytes.Replace(snippet, old, new, -1)
+
+	// convert name to AppArmor dbus interface (eg, 'org.foo' to 'org.foo{,.*}')
+	var ifaceBuf bytes.Buffer
+	ifaceBuf.WriteString(`"`)
+	ifaceBuf.WriteString(name)
+	ifaceBuf.WriteString(`{,.*}"`)
```

Comment:
To be clear, we do want to make it configurable-- at some point. There is an urgency to this PR as it is blocking GNOME and KDE devs from targeting stable. This PR specifically addresses that use case. Other use cases can come in other PRs.

As for Properties, this already handled by the path rule without an interface:
```
# allow connected snaps to all interfaces via ###DBUS_PATH### (eg,
# org.freedesktop.*, org.gtk.Application, etc) to allow full integration with
# connected snaps.
dbus (receive, send)
    bus=###DBUS_BUS###
    path=###DBUS_PATH###
    peer=(label=###SLOT_SECURITY_TAGS###),
```

This was a tested condition for the leaf applications we are trying to address.